constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ViolationsOfSpans.java

```diff
@@ -0,0 +1,62 @@
+package gov.nasa.jpl.aerie.constraints.tree;
+
+import gov.nasa.jpl.aerie.constraints.model.EvaluationEnvironment;
+import gov.nasa.jpl.aerie.constraints.model.SimulationResults;
+import gov.nasa.jpl.aerie.constraints.model.Violation;
+import gov.nasa.jpl.aerie.constraints.time.Interval;
+import gov.nasa.jpl.aerie.constraints.time.Spans;
+import gov.nasa.jpl.aerie.constraints.time.Windows;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+public final class ViolationsOfSpans implements Expression<List<Violation>> {
+  public final Expression<Spans> expression;
+
+  public ViolationsOfSpans(Expression<Spans> expression) {
+    this.expression = expression;
+  }
+
+
+  @Override
+  public List<Violation> evaluate(SimulationResults results, final Interval bounds, EvaluationEnvironment environment) {
+    final var satisfiedWindows = this.expression.evaluate(results, bounds, environment);
+
+    final var activityIds = new ArrayList<Long>();
+    final var spansList = new ArrayList<Spans>();
+    satisfiedWindows.iterator().forEachRemaining(element -> {
+      if(element.value().isPresent()) {
+        activityIds.add(element.value().get().id());
+      }
+      spansList.add(new Windows(element.interval(), true).not().select(bounds).intoSpans());
+    });
+    final var allIntervals = new ArrayList<Interval>();
+    spansList.iterator().forEachRemaining(a-> a.iterator().forEachRemaining(b -> allIntervals.add(b.interval())));
+    return List.of(new Violation(activityIds, new ArrayList<>(), allIntervals));
+  }
```

Comment:
I thought since we reduced scope by leaving `Constraint.ForEachActivity`, we wouldn't be needing this anymore.

BTW this: `new Windows(element.interval(), true).not().select(bounds).intoSpans()` should always produce an exception because the `Windows(interval, bool)` constructor makes everything a gap outside of the interval, and `select(bounds)` will set gaps outside the bounds anyway; then `intoSpans` throws an exception because it can't be called on windows with gaps.

Reply:
I'm seeing a bit of an issue coming up everywhere in the constraints DSL because of this. We eagerly restrict and truncate evaluation to the plan bounds, but this is problematic for operations that don't like gaps, or for `shiftBy` which can lose data by shifting it outside of the bounds (and back).

My solution is I'd like to remove all the `select` operations in the AST nodes except for the top-level constraint nodes.

Reply:
I can remove it. 
Regarding your other comment : what nodes are top-level ? We can use any node and the contract of `evaluate` is that if we pass bounds, we get results only inside these bounds, no ?

Reply:
I'm proposing that we should remove that contract. The `ViolationsOfWindows` and `ForEachActivityViolations` nodes can restrict the bounds, because we know no more operations will be done. Similarly for scheduling, the consumer of the windows objects in the scheduler can decide whether or not to truncate the result to the plan.

But it doesn't make sense to truncate the result of `profile.equals(5)` to the plan bounds immediately, because if it was an external profile, there might be data outside of the plan that will be moved in by a `shiftBy`.

Reply:
Maybe you can make a ticket or a discussion on the slack and we can discuss about this outside of this PR ?

Reply:
Ok, yes this is worthy of a github discussion.

Reply:
FWIW, Joel and I discussed a potential future direction for the `IntervalMap` data structure that would compute lazily / on-demand rather than eagerly (and potentially wastefully) computing across the whole timeline on every operation on a profile. I think that would fit quite snugly into the proposed GitHub discussion.